edgelet/edgelet-docker/src/image_prune_data.rs

```diff
@@ -0,0 +1,652 @@
+            ),
+        };
+
+        let fp: PathBuf = homedir.join(IMAGE_USE_FILENAME);
+        let tmp_fp: PathBuf = homedir.join(TMP_FILENAME);
+
+        let image_use_filepath = fp
+            .to_str()
+            .ok_or_else(|| Error::FilepathCreationError(IMAGE_USE_FILENAME.into()))?;
+        let tmp_filepath = tmp_fp
+            .to_str()
+            .ok_or_else(|| Error::FilepathCreationError(IMAGE_USE_FILENAME.into()))?;
+
+        Ok(Self {
+            inner: Arc::new(Mutex::new(ImagePruneInner {
+                image_use_filepath: image_use_filepath.to_string(),
+                tmp_filepath: tmp_filepath.to_string(),
+                settings,
+            })),
+        })
+    }
+
+    /// This method takes the `image_id` and adds (if the image is new) OR updates the last-used timestamp associated
+    /// with this `image_id`. This state is maintained for use during image garbage collection.
+    /// This method is (currently) called whenever a new image is pulled, a container is created, or when a conatiner is removed.
+    pub fn record_image_use_timestamp(&self, image_id: &str) -> Result<(), Error> {
+        let guard = self
+            .inner
+            .lock()
+            .map_err(|e| Error::LockError(e.to_string()))?;
+
+        // read persistence file into in-mem map
+        // this map now contains all images deployed to the device (through an IoT Edge deployment)
+
+        let mut image_map = match get_images_with_timestamp(guard.image_use_filepath.clone()) {
+            Ok(map) => map,
+            Err(e) => {
+                drop(guard);
+                log::warn!("Could not read auto-prune data. Image garbage collection did not prune any images. Error: {}", e);
+                return Ok(());
+            }
+        };
+
+        let current_time = std::time::SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Could not get EPOCH time");
+
+        image_map.insert(image_id.to_string(), current_time);
+
+        // write entries back to file
+        let _res = write_images_with_timestamp(
+            &image_map,
+            guard.tmp_filepath.clone(),
+            guard.image_use_filepath.clone(),
+        );
+
+        drop(guard);
+
+        Ok(())
+    }
+
+    /// This method is called during image garbage collection. It returns a map of images that
+    /// will be deleted by the image garbage collector.
+    /// The `in_use_image_ids` is a set of image IDs currently being used on the device.
+    pub fn prune_images_from_file(
+        &self,
+        in_use_image_ids: HashSet<String>,
+    ) -> Result<HashMap<String, Duration>, Error> {
+        let guard = self
+            .inner
+            .lock()
+            .map_err(|e| Error::LockError(e.to_string()))?;
+
+        let settings = guard.settings.clone();
+
+        // if pruning is disabled then shouldn't remove anything
+        if !settings.is_enabled() {
+            return Ok(HashMap::new());
```

Comment:
This is a bug - mutex not released.

This check should probably go here: https://github.com/gauravIoTEdge/iotedge/blob/image-garbage-collection/edgelet/aziot-edged/src/image_gc.rs#L137